File: sw-python/zoomg3xn.py
import rtmidi
import threading
from logger import Logger
from enum import Enum

class ZoomG3Xn:

    class _MESSAGE_TYPES(Enum):
        EDITOR_MODE_ON = 1
        EDITOR_MODE_OFF = 2
        TOGGLE = 3

    class _MESSAGE_PARAMS(Enum):
        CHAIN_POSITION = 7
        STATE = 9

    _MESSAGE = {
        _MESSAGE_TYPES.EDITOR_MODE_ON: [0xf0, 0x52, 0x00, 0x6e, 0x50, 0xf7], 
        _MESSAGE_TYPES.EDITOR_MODE_OFF: [0xf0, 0x52, 0x00, 0x6e, 0x51, 0xf7], 
        _MESSAGE_TYPES.TOGGLE: [0xf0, 0x52, 0x00, 0x6e, 0x64, 0x03, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf7] # idx 7 is position in chain
    }

    def __init__(self, port_name = "ZOOM"):
        self.log = Logger('zoomg3xn')
        self.log.warning("Opening MIDI connection to pedalboard...")
        # Create midi interfaces
        self.midiin = rtmidi.MidiIn()
        self.midiout =rtmidi.MidiOut()
        # Allow sysex communication
        self.midiin.ignore_types(sysex=False)

        in_ports = self.midiin.get_ports()
        out_ports = self.midiout.get_ports()

        if not in_ports or not out_ports:
            raise RuntimeError("No midi ports")
        for idx, p in enumerate(in_ports):
            if port_name in p:
                self.midiin.open_port(idx)
                self.log.info("Midi In port ready")
                break
        for idx, p in enumerate(out_ports):
            if port_name in p:
                self.midiout.open_port(idx)
                self.log.info("Midi Out port ready")
                break
        if not self.midiin.is_port_open() or not self.midiout.is_port_open():
            raise RuntimeError("A port wasn't opened")
        
        self.on_receive = None
        self.thread = threading.Thread(target=self._loop, daemon=True)
        self.log.info("Pedalboard connected")


    def _loop(self):
        self.midiout.send_message(self._MESSAGE[self._MESSAGE_TYPES.EDITOR_MODE_ON]) # Edit mode
        self.log.info("Pedalboard thread started.")

        while True:
            if not self.midiin.is_port_open() or not self.midiout.is_port_open():
                self.log.error("Midi ports disconnected, stopping thread")
                break
            msg = self.midiin.get_message()
            if msg:
                bytes_res = bytes(msg[0]).hex()
                self._parse_message(bytes_res)

    def start(self):
        self.log.warning("Starting pedalboard thread...")
        self.thread.start()

    def toggleEffect(self, effect_num, state):
        # Send sysex to activate/deactivate effect
        _msg = self._MESSAGE[self._MESSAGE_TYPES.TOGGLE]
        _msg[self._MESSAGE_PARAMS.CHAIN_POSITION.value] = effect_num
        _msg[self._MESSAGE_PARAMS.STATE.value] = state
        self.midiout.send_message(_msg)

        self.log.info("Toggled {}: {}".format(effect_num, 'ON' if state == 1 else 'OFF'))

    def _parse_message(self, msg: str, **kwargs):
        # parse the midi/sysex message and translate it to <command>:<number> format
        if msg.startswith('f052006e6403'):
            effect_num = int(msg[14:16], 16)
            state = int(msg[18:20], 16)
            _parsed = "{}:{}".format("on" if state == 1 else "off", effect_num)
            self.log.info("Got midi message, {}".format(_parsed))
            if self.on_receive:
                self.on_receive(_parsed, **kwargs)

    def thread_alive(self):
        return self.thread.is_alive()
        

File: sw-cpp/zoomg3xn.cpp
#include "zoomg3xn.hpp"

enum MESSAGE_PARAMS {
    CHAIN_POSITION = 7,
    STATE = 9
};

const unsigned char MESSAGE_EDITOR_MODE_ON[] = {0xf0, 0x52, 0x00, 0x6e, 0x50, 0xf7};
const unsigned char MESSAGE_EDITOR_MODE_OFF[] = {0xf0, 0x52, 0x00, 0x6e, 0x51, 0xf7};
unsigned char MESSAGE_TOGGLE[] = {0xf0, 0x52, 0x00, 0x6e, 0x64, 0x03, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xf7}; // params at enum positions

ZoomG3Xn::ZoomG3Xn(std::string port_name, RtMidiIn::RtMidiCallback on_message_callback) {
    log = new Logger("zoomg3xn");
    try {
        midiin = new RtMidiIn();
        midiout = new RtMidiOut();
    } catch ( RtMidiError &error) {
        log->error(error.getMessage());
        exit( EXIT_FAILURE );
    }

    // Setup ignoreTypes
    midiin->ignoreTypes(false, true, true);

    log->info("Starting MIDI ports configurations");

    unsigned int portCount = midiin->getPortCount();
    if (portCount == 0) {
        log->error("No MIDI input ports");
        exit( EXIT_FAILURE );
    }
    for (unsigned int i = 0; i < portCount; i++) {
        try {
            std::string portName = midiin->getPortName(i);
            if (portName.find(port_name) != std::string::npos) {
                midiin->openPort(i);
                midiin->setCallback(on_message_callback);
                break;
            }
        } catch ( RtMidiError &error ) {
            log->error(error.getMessage());
            exit( EXIT_FAILURE );
        }
    }
    if (!midiin->isPortOpen()) {
        log->error("MIDI input port is not open");
        exit( EXIT_FAILURE );
    }
    log->info("MIDI input port ready");

    portCount = midiout->getPortCount();
    if (portCount == 0) {
        log->error("No MIDI output ports");
        exit( EXIT_FAILURE );
    }
    for (unsigned int i = 0; i < portCount; i++) {
        try {
            std::string portName = midiout->getPortName(i);
            if (portName.find(port_name) != std::string::npos) {
                midiout->openPort(i);
                break;
            }
        } catch ( RtMidiError &error ) {
            log->error(error.getMessage());
            exit( EXIT_FAILURE );
        }
    }
    if (!midiout->isPortOpen()) {
        log->error("MIDI output port is not open");
        exit( EXIT_FAILURE );
    }
    log->info("MIDI output port ready");

    std::vector<unsigned char> start(MESSAGE_EDITOR_MODE_ON, MESSAGE_EDITOR_MODE_ON + (sizeof(MESSAGE_EDITOR_MODE_ON)/sizeof(unsigned char)));
    try {
        midiout->sendMessage(&start);
    } catch ( RtMidiError &error ) {
        log->error(error.getMessage());
        exit( EXIT_FAILURE );
    }
    log->info("Zoom pedalboard connected");
}

ZoomG3Xn::~ZoomG3Xn() {
    midiin->closePort();
    midiout->closePort();

    delete log;
    delete midiin;
    delete midiout;
}

void ZoomG3Xn::toggle_effect(int num_effect, bool state) {
    std::vector<unsigned char> _msg(MESSAGE_TOGGLE, MESSAGE_TOGGLE + (sizeof(MESSAGE_TOGGLE)/sizeof(unsigned char)));

    _msg[MESSAGE_PARAMS::CHAIN_POSITION] = (unsigned char)num_effect; 
    _msg[MESSAGE_PARAMS::STATE] = state ? 0x1 : 0x0;

    try {
        midiout->sendMessage(&_msg);
    } catch ( RtMidiError &error ) {
        log->error(error.getMessage());
        if (!midiout->isPortOpen())
            exit( EXIT_FAILURE );
    }
}
